xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-webjar/src/main/webjar/persistent-configuration.vue

```diff
@@ -0,0 +1,229 @@
+          (filterDescriptor.operators || []).forEach(function (operator) {
+            if (operators.indexOf(operator.id) === -1) {
+              operators.push(operator.id);
+            }
+          });
+        });
+        return ["oneOf"].concat(operators);
+      },
+
+      encodingSpecsCurrentLayoutId: function () {
+        return ["oneOf"].concat(this.logic.getLayoutIds());
+      },
+
+      encodingSpecsV1: function () {
+        return {
+          filters: ["array", {
+            property: this.encodingSpecsProperties,
+            matchAll: ["boolean"],
+            constrains: ["array", {
+              operator: this.encodingSpecsFilterOperators,
+              value: ["varchar"],
+            }],
+          }],
+          sort: ["array", {
+            property: this.encodingSpecsProperties,
+            descending: ["boolean"],
+          }],
+          offset: ["integer"],
+          limit: ["integer"],
+          currentLayoutId: this.encodingSpecsCurrentLayoutId,
+          hiddenProperties: ["array", this.encodingSpecsProperties],
+          propertyOrder: ["array", this.encodingSpecsProperties],
+        };
+      },
+
+      coders: function () {
+        return [
+          u.fromJson(1, this.encodingSpecsV1),
+        ];
+      },
+
+      dataToSave: function () {
+        var dataToSave = {};
+        for (var key in this.encodingSpecsV1) {
+          dataToSave[key] = this.getDataParentObject(key)[key];
+        }
+        return dataToSave;
+      },
+
+      encodedConfig: function () {
+        return this.encodeConfig(this.dataToSave);
+      },
+
+    },
+
+    watch: {
+      encodedConfig: function () {
+        if (this.encodedConfig) {
+          this.saveConfig();
+        } else {
+          this.deleteConfig();
+        }
+      },
+    },
+
+
+    methods: {
+
+      getDataParentObject: function (key) {
+        if (this.data.query.hasOwnProperty(key)) {
+          return this.data.query;
+        } else {
+          return this.logic;
+        }
+      },
+
+      encodeConfig: function (config) {
+        try {
+          var encoded = u.encode(this.coders[this.coders.length -1], config);
+          var compressed = LZString.compressToEncodedURIComponent(encoded);
+          return compressed;
+        } catch (err) {
+          console.warn(err);
+          return "";
+        }
+      },
+
+      decodeConfig: function (encodedConfig) {
+        if (!encodedConfig) { return; }
+        var uncompressed = LZString.decompressFromEncodedURIComponent(encodedConfig);
+        if (!uncompressed) { return; }
+        try {
+          var decoded = u.decode(this.coders, uncompressed);
+          return decoded;
+        } catch (err) {
+          console.warn(err);
+        }
+      },
+
+      saveConfig: function () {
+        // url hash
+        if (this.urlHash) {
+          var url = new URL(window.location);
+          url.searchParams.set("livedata-config", this.encodedConfig);
```

Comment:
The server doesn't receive the hash on page reload but it does receive the search parameters so it could react differently based on the received search parameters (query string). This means that the user might get a different thing on page reload than what they expect. The chances are low but it's possible.

On the same subject, what happens if you have multiple live data instances on the page? Can they have different configuration stored on the URL / local storage?

Reply:
> what happens if you have multiple live data instances on the page? Can they have different configuration stored on the URL / local storage?

It's not done yet. Is there a way to identify a component to be unique, like with its macro id (if there is one)?

Reply:
The user that adds the live data macro can specify an id that ends up in the live data configuration, at the top level:

```
{
  "id": "foo",
  "query": {...},
  ...
}
```

You could use that but we need to decide what to do when the user doesn't specify an id:
* either don't store the state in the URL
* or generate an automatic id, either on the client or on the server.

On the client side, you could maybe keep a count of the live data instances created and use that as an id when there's no id specified by the user. WDYT?

Reply:
Great, I already used the id from `query.source` in an upcoming commit, but I wasn't sure it was the right way to do it. 
I think the most robust way is to generate automatically an id on the server if the user does not, so that the client always receive an id in any case. (also why does the user has the possibility to specify himself an id? why it is not automatically generated all the time?)

> On the client side, you could maybe keep a count of the live data instances created and use that as an id when there's no id specified by the user. WDYT?

The problem of this solution I think is that if a LD is added or removed from the page (for example a LD is added before the existing one), then the generated id will no longer match the LDs on the page.

Reply:
> why does the user has the possibility to specify himself an id?

Because the user (developer actually) needs to **interact** with the live data instance and / or **customize** it at runtime. This is actually quite common. The developer that inserted the live data instance needs to be able to identify and access it from their own JavaScript code, and also to customize its styles from CSS. From JavaScript the basic need is to listen to events triggered by a specific live data instance or to use directly its API to reload the data, e.g. when the data on the server has been updated by something else.

> I think the most robust way is to generate automatically an id on the server if the user does not

I propose to disable the URL hashing if the live data id is missing. We can add the automatic id generation on the server afterwards.

> I already used the id from query.source in an upcoming commit

The source id and the live data id are different. You can have two live data instances that use the same source but otherwise have different configuration (e.g. different filters). So I propose to have an ``id`` field on the root of the live data configuration.

> The problem of this solution I think is that if a LD is added or removed from the page (for example a LD is added before the existing one), then the generated id will no longer match the LDs on the page.

Indeed.

Reply:
> I propose to disable the URL hashing if the live data id is missing. We can add the automatic id generation on the server afterwards.

Okay, let's do like that for know! :)